src/miral/keymap.cpp

```diff
@@ -244,3 +245,166 @@ void miral::Keymap::set_keymap(std::string const& keymap)
 {
     self->set_keymap(keymap);
 }
+
+miral::Keymap::Keymap(std::unique_ptr<Self>&& self) :
+    self{std::move(self)}
+{
+}
+
+namespace
+{
+using namespace miral::gdbus;
+
+char const* const interface_name = "org.freedesktop.locale1";
+char const* const bus_name = interface_name;
+char const* const sender = interface_name;
+char const* const object_path = "/org/freedesktop/locale1";
+char const* const properties_interface = "org.freedesktop.DBus.Properties";
+char const* const signal_name = "PropertiesChanged";
+
+auto read_entry(Connection const& connection, char const* entry) -> std::optional<std::string>
+{
+    static char const* const method_name = "Get";
+
+    GError* error = nullptr;
+
+    if (Variant const result{g_dbus_connection_call_sync(connection,
+                                                        bus_name,
+                                                        object_path,
+                                                        properties_interface,
+                                                        method_name,
+                                                        g_variant_new("(ss)", interface_name, entry),
+                                                        nullptr,
+                                                        G_DBUS_CALL_FLAGS_NONE,
+                                                        G_MAXINT,
+                                                        nullptr,
+                                                        &error)})
+    {
+        Variant const unwrap{g_variant_get_child_value(result, 0)};
+        Variant const unwrap2{g_variant_get_child_value(unwrap, 0)};
+
+        if (g_variant_is_of_type(unwrap2, G_VARIANT_TYPE_STRING))
+        {
+            gsize len = 0;
+            return g_variant_get_string(unwrap2, &len);
+        }
+    }
+
+    if (error)
+    {
+        mir::log_info("Dbus error=%s, dest=%s, object_path=%s, properties_interface=%s, method_name=%s, interface_name=%s",
+                      error->message, bus_name, object_path, properties_interface, method_name, interface_name);
+        g_error_free(error);
+    }
+
+    return std::nullopt;
+}
+
+auto read_keymap(Connection const& connection) -> std::optional<std::string>
+{
+    if (auto const layout = read_entry(connection, "X11Layout"))
```

Comment:
This is three blocking round-trips to the D-Bus server - you could do this in one with GetAll. Perhaps add a comment so it could be improved in the future.